torch/csrc/jit/serialization/export_bytecode.cpp

```diff
@@ -1,11 +1,331 @@
 
 #include <torch/csrc/jit/runtime/instruction.h>
 #include <torch/csrc/jit/serialization/export.h>
 
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/api/function_impl.h>
+#include <torch/csrc/jit/api/method.h>
+#include <torch/csrc/jit/backends/backend_debug_handler.h>
+#include <torch/csrc/jit/backends/backend_debug_info.h>
+#include <torch/csrc/jit/frontend/source_range.h>
+#include <torch/csrc/jit/ir/attributes.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/type_hashing.h>
+#include <torch/csrc/jit/mobile/function.h>
+#include <torch/csrc/jit/mobile/interpreter.h>
+#include <torch/csrc/jit/mobile/method.h>
+#include <torch/csrc/jit/mobile/module.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/runtime/instruction.h>
+#include <torch/csrc/jit/serialization/callstack_debug_info_serialization.h>
+#include <torch/csrc/jit/serialization/export_bytecode.h>
+#include <torch/csrc/jit/serialization/import_export_constants.h>
+#include <torch/csrc/jit/serialization/import_export_functions.h>
+#include <torch/csrc/jit/serialization/import_export_helpers.h>
+#include <torch/csrc/jit/serialization/pickle.h>
+#include <torch/csrc/jit/serialization/python_print.h>
+#include <torch/csrc/jit/serialization/source_range_serialization.h>
+#include <torch/csrc/jit/serialization/type_name_uniquer.h>
+
+#include <caffe2/serialize/inline_container.h>
+
 namespace torch {
 namespace jit {
 
+std::vector<Method> gatherMethodsToExport(
+    const Module& module,
+    bool incl_setstate,
+    bool incl_interface_call) {
+  std::vector<Method> methods_to_export;
+  std::deque<Method> stack;
+
+  for (const auto& method : module.get_methods()) {
+    stack.emplace_back(method);
+  }
+
+  if (incl_setstate) {
+    // Use DFS on IValue's to traverse dependencies of module._ivalue and
+    // add all setstate/getstates to initial stack.
+    std::vector<ObjectPtr> ivalue_stack;
+    ivalue_stack.push_back(module._ivalue());
+    while (!ivalue_stack.empty()) {
+      ObjectPtr cur = ivalue_stack.back();
+      ivalue_stack.pop_back();
+      auto type = cur->type();
+      Function* setstate = type->findMethod("__setstate__");
+      Function* getstate = type->findMethod("__getstate__");
+      if (getstate && setstate) {
+        if (setstate->isGraphFunction()) {
+          stack.emplace_back(cur, setstate);
+        }
+        if (getstate->isGraphFunction()) {
+          stack.emplace_back(cur, getstate);
+        }
+      } else {
+        for (size_t i = 0, n = type->numAttributes(); i < n; ++i) {
+          IValue field = cur->getSlot(i);
+          if (field.isObject()) {
+            ivalue_stack.push_back(field.toObject());
+          }
+        }
+      }
+    }
+  }
+
+  // After getting the initial set of functions to export;
+  // recursively search for those that these function calls.
+  // BFS to cover ALL methods of submodules that are used
+  std::set<std::pair<std::string, Function*>> visited;
```

Comment:
I like the idea that gatherMethodsToExport and compileGraphToMobileCode is separated. Here the problem is to find all interface calls, we need to inline the graph for once, and dropped the inlined graph, then when we actually compile it, we inline it again with the same result. This is probably fine, but I think in general we should only do the needed work exactly once. That's one of the reason why ExportedBytecodeSet is there, also to cache inlined graph.

I'd suggest we reuse and move out the private `exportFunction` from `export_module.cpp` which should already handle these things. More specially, my "ideal" gatherMethodsToExport is like:
```
def gatherMethodsToExport() {
  BytecodeExportSet exportSet;
  for (method : module.methods) {
    exportFunction(exportSet, method);
  }
}
```
Now if you don't like recursive `exportFunction`, another option is still keeping BFS, but at least leverage cached graphs, like:
```
def gatherMethodsToExport() {
  BytecodeExportSet exportSet;
  // BFS loop
  while (...) {
    // ...
    exportSet.add(stack_top, cached_graph);
    // ...
  }
}
```